proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,393 @@
+
+Effort will be made to include them if possible within the proposed timeline, and the project may be extended at the discretion of the Haskell Foundation in order to ensure their completion.
+
+
+# Appendix A - Commercial Incentive Misalignment
+
+There is no open, community-driven, community-owned cryptography solution, and it is unlikely that one will be developed without community investment. Commercial development is unlikely to provide us with a solution, due to the issue of ownership of the resulting ecosystem, and the resulting incentive misalignment:
+
+- Cryptography and security are preventative measures and a cost center for most companies
+- Companies for which cryptography / security are a product have a strong incentive to marshal users into a walled garden in the name of both security and profit
+- Developing an ecosystem is a significant up-front investment, and so investors may expect ownership of the ecosystem in order to recoup their cost
+- There is an incentive to wait for someone else to develop an open ecosystem, instead of expending resources yourself to do it
+
+As such, commercial or for-profit approaches are incentivized to either invest in a private ecosystem, or invest in an established open ecosystem, but are perversely incentivized against building a new open ecosystem.
+
+# Appendix B - Lack of ByteString `Bits` instance
+
+Essentially, `Bits`` (and the fixed-width `FiniteBits``) is a class that combines two concepts:
+
+    `Boolean / Heyting`` algebra, for which there is no concept of any 'internal bits' or structure and you are just operating on some object as a whole
+    Things that are contructed from / encoded into a set of 'internal bits' using some encoding, for which individual bits are indexable / addressible, and thus individually boolean-operable. There is also a link here to indexing / representable, but that is getting way out of scope.
+
+Notably, the documentation itself further states that "The `Bits` class defines bitwise operations over integral types.". Technically this means that only things that are integer numbers should be `Bits` and `FiniteBits`, but Godel numbering rears its ugly head, and that just adds the question of whether `Bits`'s boolean operations are intended to apply to an object itself or its encoding, and as a result `Bits` and `FiniteBits` are restricted to objects for which the boolean operations are isomorphic to boolean operations over its encoding.
+
+Given this understanding, `ByteString` does not have a canonical `Bits` instance because it has no canonical integer representation - a given bytestring could represent many such integers, whether it be big endian or little endian, it could have a sign bit, or be two's complement. Its bits might not be contiguous or even ordered. It could easily satisfy a `Boolean` instance, but the issue of indexing individual bits and integral type requirement interferes with a full instance of `Bits`.
```

Comment:
In my CLC hat I'm hugely sceptical about introducing `class Boolean` as a superclass of `Bits`. I mean, I absolutely agree with the motivation, but we just don't have a good migration story for changing type class hierarchy. 

One solution would be to resurrect default superclasses proposal. 

I recall that `bitvec` (which is morally a `ByteString`) has an audacity to define `instance Bits`. Is it if any help? Could we just wrap `ByteString` into a newtype and define whatever is missing?

Reply:
The proposal for `class Boolean` spawns from the same work as but is not a part of this proposal. David asked for clarification / a link to the topic, so I added an addendum just in case but it can be ignored entirely.

Re: `bitvec` I already have a lightweight class for sized byte vectors in `crypto-schemes`, but it is mostly for type safety and `bitvec` may be faster. I'll definitely be keeping that in mind.